rfcs/taskforce/0001-buildkiteci.md

```diff
@@ -0,0 +1,215 @@
+
+Engineers who are iterating on build rules can run nodes locally or spin up instances to test things using the buildkite CLI
+
+### Description of pipelines
+
+Pipeline are configured with Dhall rather than jinja-templated yaml. Dynamic pipelines support this as a first-class citizen, we do not require running the equivalent "`[test.py](http://test.py)`" before pushing. I have already implemented this on a [side project](https://github.com/bkase/gameboy/blob/master/.buildkite/pipeline.yaml).
+
+**General Pipeline Writing Things**
+
+* Use the [BuildKite monorepo plugin](https://github.com/chronotc/monorepo-diff-buildkite-plugin) for granular test execution.
+* Explicitly opt-in to the dependency graph. Don't support the "wait" instruction that blocks all parallel work.
+
+Assuming this proposal is accepted, more details will follow on how to structure the Dhall descriptions, but here is a rough cut of what it could look like:
+
+```haskell
+-- This is a definition of the package-deb step for Coda
+-- Optional decorations stack on top of the required bits
+-- This step will rerun twice before giving up
+-- It will upload artifacts under /tmp/artifacts/** to google cloud
+-- It depends on coda-build, genesis-ledger, and libp2p-helper steps
+-- Then the 4 required fields are given: a nice label for the UI, the
+--   command to run, a machine-readable key for further dependency graph
+--   manipulation, and a list of globs for granular test execution.
+-- Note: By making dirtyWhen a _required_ field, we set ourselves up for
+--       a world in which CI doesn't run on docs changes
+let packageDebStep =
+   decorate3
+	   (withFlaky 2)
+	   (withArtifacts "/tmp/artifacts/**")
+		 (withDeps [ "coda-build", "genesis-ledger", "libp2p-helper" ])
+	   Command::{
+	     label = ":box: Package deb",
+	     command = "make deb"
+	     key = "package-deb",
+	     dirtyWhen = [ "src/**" ]
+	   }
+```
+
+### Workflow of adding/changing/debugging pipelines
+
+Run buildkite agents locally or on manually spun up compute instances to test rules by targeting specific agents with tags and using the CLI [https://github.com/buildkite/cli](https://github.com/buildkite/cli)
+
+Liberal use of artifact uploading/downloading to keep build re-playable from different steps
+
+## This design solves the above issues
+
+P1 → self-hosted on GKE
+
+P2.1.1 → Parallel by default; opt-in to dependencies
+
+P2.1.2 → Rely on artifacts in gcloud which should be very quick to upload and download from within Gcloud infrastructure
+
+P2.2 → We can choose whatever machines we want
+
+P2.3 → We use dynamic pipelines to expand work based on files touched
+
+P3 → Manual adhoc template expansion is replaced with *automatic* expansion using Dhall
```

Comment:
We still need to render this into a config for buildkite, right? If not, can you describe what "automatic expansion" using Dhall means somewhere in this RFC?

Reply:
Nope! The dynamic pipelines allow for the expansion to happen at Ci execution time. I already have this working on my Gameboy project:

https://github.com/bkase/gameboy/tree/master/.buildkite

You just need a small bootstrap pipeline.yaml that injects the pipeline.dhall

I will expand on this in the RFC to make it more clear

Reply:
That is nice. Can we have a pre-commit git-hook that checks that the dhall is renderable before pushes? Or is that too difficult to consider?

Reply:
> That is nice. Can we have a pre-commit git-hook that checks that the dhall is renderable before pushes? Or is that too difficult to consider?

This should be easy `for f in $(dhall-files); do dhall <<< "./${f}" || fail "file $f is not renderable"; done`